src/util/XDRStream.h

```diff
@@ -118,10 +136,78 @@ class XDRInputFileStream
         {
             throw xdr::xdr_runtime_error("malformed XDR file");
         }
+
         xdr::xdr_get g(mBuf.data(), mBuf.data() + sz);
         xdr::xdr_argpack_archive(g, out);
         return true;
     }
+
+    template <typename T>
+    bool
+    readPage(T& out, LedgerKey const& key, size_t pageSize)
+    {
+        ZoneScoped;
+        if (mBuf.size() < pageSize)
+        {
+            mBuf.resize(pageSize);
+        }
+
+        if (!mIn.read(mBuf.data(), pageSize))
+        {
+            if (mIn.eof())
+            {
+                mIn.clear(std::ios_base::eofbit);
+            }
+            else
+            {
+                return false;
+            }
+        }
+
+        auto bytesToRead = mIn.gcount();
+        char* curr = mBuf.data();
+        for (std::streamsize read = 0; read < bytesToRead;)
+        {
+            ZoneNamedN(__unpack, "xdr_unpack_entry", true);
+
+            uint32_t sz = getXDRSize(curr);
+            read += sizeof(sz);
+            curr += sizeof(sz);
+
+            // If entry continues past end of page
+            if (read + sz > bytesToRead)
+            {
+                if (mBuf.size() < sz)
+                {
+                    mBuf.resize(sz);
+                }
+
+                // Move undecoded bytes to front of buffer and read in the rest
+                // of the entry
+                auto bytesRemaining = bytesToRead - read;
+                memmove(mBuf.data(), curr, bytesRemaining);
```

Comment:
I'm not really sure why the `memmove` is happening here, nor why it is enough to make space for the subsequent read to succeed (I think the previous resize likely _shrank_ the buffer?); at this point the offset and pointer arithmetic becomes too unclear for me to follow.

I think it would make more sense to expand the buffer to accommodate the remainder, rather than maybe-shrink or maybe-expand it to cover the current record. And fill only that remainder piece. I'll post a change that does this (and fixes the other issues).

Reply:
I was using the `memove` to avoid having to resize the buffer and instead just throw out all the stuff we already read, move the unread portion to the front of the buffer, then read in the rest of the data that we need. I like yours a LOT better.